[MI-3766] Fixed the issue of mentions not working properly after using certificate based subscriptions

manojmalik20 opened the pull request:
#### Summary
Fixed the issue of mentions not working properly after using certificate based subscriptions

#### Ticket Link
Fixes #391

lieut-data reviewed (changes requested):
Thanks for driving this, @manojmalik20! I understand these changes are the result of payloads received while certificate based subscriptions are enabled. Was this also broken previously? Are these changes compatible with messages received when certificate based subscriptions are disabled?

Can we add unit tests to the converters to showcase the cases we're fixing?

manojmalik20 commented:
Hey @lieut-data , 

Yes, the changes are compatible with messages received when certificate based subscriptions are disabled. Actually, we misinterpreted the whole thing. I tested the original code with certifcate based subscriptions disabled and I found that the mentions were broken there too. Recently, the PR for certificate based subscriptions was merged so we thought that the mentions were breaking due to that, but actually MS Graph has changed the way they return mentions in a chat message recently. Due to that, our logic for parsing the mentions from MS Teams and converting them into Mattermost mentions was breaking. If we see the documentation for the [Get ChatMessage](https://learn.microsoft.com/en-us/graph/api/chatmessage-get?view=graph-rest-1.0&tabs=http), the documentation was updated on November 18th, so I think that's when the mentions started breaking.
Can we verify this in the ServiceNow test or prod environment?

I've fixed this issue now and updated the PR title.

ayusht2810 commented on server/handlers/converters.go:
```diff
@@ -97,8 +98,12 @@ func (ah *ActivityHandler) handleMentions(msg *clientmodels.Message) string {
-		case mention.UserID != "":
+		if mention.UserID != "" {
+			if userOrConversationIDVsNames[mention.UserID] != mention.MentionedText {
+				msg.Text = strings.Replace(msg.Text, fmt.Sprintf("&nbsp;<at id=\"%s\">%s</at>", fmt.Sprint(mention.ID), mention.MentionedText), "", 1)
```
why are we also replacing the initial space here?

manojmalik20 replied:
The mention `Dummy User` on Teams should be converted to `@dummyuser` on Mattermost without a trailing space. This is done to remove that trailing space.

ayusht2810 commented on server/handlers/converters.go:
```diff
@@ -75,19 +75,20 @@ func (ah *ActivityHandler) msgToPost(channelID, senderID string, msg *clientmode
 
 func (ah *ActivityHandler) handleMentions(msg *clientmodels.Message) string {
-	userIDVsNames := make(map[string]string)
+	// This map is used to store user or conversation IDs vs the first word of the multi-word mention
```
is it the first or last word?

manojmalik20 commented:
Hey @lieut-data , 

Yes, the changes are compatible with messages received when certificate based subscriptions are disabled. Actually, we misinterpreted the whole thing. I tested the original code with certifcate based subscriptions disabled and I found that the mentions were broken there too. Recently, the PR for certificate based subscriptions was merged so we thought that the mentions were breaking due to that, but actually MS Graph has changed the way they return mentions in a chat message recently. Due to that, our logic for parsing the mentions from MS Teams and converting them into Mattermost mentions was breaking. If we see the documentation for the [Get ChatMessage](https://learn.microsoft.com/en-us/graph/api/chatmessage-get?view=graph-rest-1.0&tabs=http), the documentation was updated on November 18th, so I think that's when the mentions started breaking.
Can we verify this in the Hercules test or prod environment?

I've fixed this issue now and updated the PR title.

lieut-data reviewed (changes requested):
Proposing we defer this from v1.4.0 and aim for the release in two weeks with v1.5.0. There's a few things I'd like to suggest we tackle here, but I'm focused elsewhere today -- will aim to provide feedback later, but for now let's hold on merging this.

lieut-data commented:
Will need to revisit this in v1.6.0 or later.
